packages/react-import-remote/src/hooks.ts

```diff
@@ -0,0 +1,100 @@
+import * as React from 'react';
+import {useIntersection} from '@shopify/react-intersection-observer';
+import {
+  DeferTiming,
+  WindowWithRequestIdleCallback,
+  RequestIdleCallbackHandle,
+} from '@shopify/async';
+import load from './load';
+
+interface Options<Imported> {
+  nonce?: string;
+  defer?: DeferTiming;
+  onError?(error: Error): void;
+  getImport(window: Window): Imported;
+  onImported?(imported: Imported): void;
+}
+
+type State = 'loading' | 'error' | 'loaded' | null;
+
+export function useImportRemote<Imported = any>(
+  source: string,
+  options: Options<Imported>,
+): {
+  state: State;
+  imported: Imported | null;
+  intersectionRef: any;
+} {
+  const [intersection, intersectionRef] = useIntersection();
```

Comment:
I forget if this one is expensive, but it might be one where it's better to only do it conditionally, but guard against them flipping between `defer` options (to prevent you from being able to actually violate the rule of hooks)